### PR TITLE
Fixed #32365 -- Made zoneinfo the default timezone implementation.

### PR DESCRIPTION
**Pull request by carltongibson**
A fresh PR following on from #14669 to clear the discussion. 

There were two points there worth bringing forward: 

* The discussion with @pganssle ref whether to have `django.utils.timezone.utc` just be `datetime.timezone.utc`. On that, whilst the `__getattr__` shim during the deprecation is more complex, I'm (still) leaning towards thinking it's cleaner for our users if that's the same type as we have elsewhere. i.e that we should always use a `zoneinfo.ZoneInfo` to avoid surprises. (I'm very open to a consensus emerging there though... 🆘)
* @smithdc1's [point about the branch coverage](https://github.com/django/django/pull/14669#discussion_r691578121) in the `pytz` case. No doubt there's a couple of those. Help spotting them appreciated. 

I **think** 🤞I've addressed the other points from the discussion. 

Then there's one broader topic to at least think about: namely, that there's a(n inevitable) potential change in behaviour here: 

* I've added a deprecation of `make_aware(..., is_dst=...) as [Paul mentioned on his earlier PR](https://github.com/django/django/pull/13877#issuecomment-758769902). 
* This kind of means (does mean?) we don't really handle ambiguous or non-existent datetimes. (…)
* In the release notes, I've pointed to [`pytz_deprecation_shim`](https://pytz-deprecation-shim.readthedocs.io/en/latest/index.html), which has lots of advice in the Migration Guide section. 
* But do we need more in Django? 

I think we need to be open to adding new API, but I'm not sure _a priori_ what's needed. I think we should just be open to adding it as issues come in over the 4.x cycle. I thought about a comment on the release notes saying _Bring us your issues_, but I didn't add that yet. `zoneinfo` is new. I think it's OK if we grow into the edge cases. I don't know that we have time for much now for 4.0 itself. 

I think it's not far off so comment appreciated. (No doubt CI blows up when I hit "Create" 😜)

### CODE REVIEW

**Comment by carltongibson** [comment]
@pganssle I don't seem to be able to request you as a reviewer (I guess since you're not in the Django org) but if you could pass an eye over it again, that would be amazing. 🥇

File: docs/releases/4.0.txt [hunk 1]
```diff
@@ -28,6 +28,47 @@ The Django 3.2.x series is the last to support Python 3.6 and 3.7.
...
+audit your code, since ``pytz`` and ``zoneinfo`` are not entirely equivalent:
+
+* The transitional :setting:`USE_DEPRECATED_PYTZ` setting is provided to
+  continue using ``pytz`` during the 4.x release cycle. This setting will be
+  removed in Django 5.0.
+
+* Usages of :func:`django.utils.timezone.activate`,
+  :func:`django.utils.timezone.override`, or the :setting:`DATABASE-TIME_ZONE`
+  setting, as well as calls to the ``pytz`` ``normalize()`` or ``localize()``
+  functions are candidate locations for review.
```
Comment (aaugustin) [inline docs/releases/4.0.txt:59]: Can we give or link to instructions about how to update calls to `normalize()` or `localize()`?

Regarding `activate()` and `override()`, I have no idea what review or update, which leads me to believe that much of our target audience doesn't know either ;-) In which circumstances do they break?

File: docs/releases/4.0.txt [hunk 2]
```diff
@@ -28,6 +28,47 @@ The Django 3.2.x series is the last to support Python 3.6 and 3.7.
...
+This is the next step in the migration from using ``pytz`` to using
+:mod:`zoneinfo`. Django 3.2 allowed the use of non-``pytz`` time zones. Django
+4.0 makes ``zoneinfo`` the default implementation. Support for ``pytz`` is now
+deprecated and will be removed in Django 5.0.
+
+:mod:`zoneinfo` is part of the Python standard library from Python 3.9. The
+``backports.zoneinfo`` package is automatically installed alongside Django if
+you are using Python 3.8.
+
+To the extent that you are always operating on aware datetimes with UTC, then
```
Comment (aaugustin) [inline docs/releases/4.0.txt:46]: I think that's exceedingly conservative. Conversions in forms and templates are also transparent. The only thing that isn't transparent is explicit operations on non-UTC timezones in the model layer.

File: docs/releases/4.0.txt (hunk 1 above)
Comment (carltongibson) [reply docs/releases/4.0.txt:59]: Thé migration guide in the `pytz_deprecation_shim` docs is a good place to link to.

File: docs/releases/4.0.txt (hunk 2 above)
Comment (carltongibson) [reply docs/releases/4.0.txt:46]: I'll rephrase. Thanks.

File: docs/topics/i18n/timezones.txt [hunk 3]
```diff
@@ -119,8 +123,8 @@ receives one, it attempts to make it aware by interpreting it in the
 :ref:`default time zone <default-current-time-zone>` and raises a warning.
 
 Unfortunately, during DST transitions, some datetimes don't exist or are
-ambiguous. In such situations, pytz_ raises an exception. That's why you should
-always create aware datetime objects when time zone support is enabled.
+ambiguous. That's why you should always create aware datetime objects when time
```
Comment (smithdc1) [inline docs/topics/i18n/timezones.txt:126]: I wonder if it's worth documenting what you mean by ambiguous. It is mentioned further down so maybe that's enough.

File: docs/topics/i18n/timezones.txt [hunk 4]
```diff
@@ -683,19 +684,14 @@ Usage
...
    In this example, the current time zone is ``"Europe/Paris"``.
 
 #. **How can I see all available time zones?**
 
-   pytz_ provides helpers_, including a list of current time zones and a list
-   of all available time zones -- some of which are only of historical
-   interest. :mod:`zoneinfo` also provides similar functionality via
-   :func:`zoneinfo.available_timezones`.
+   :func:`zoneinfo.available_timezones` provides the set of all valid keys for
+   IANA time zones available to your system. See the docs for usage
```
Comment (smithdc1) [inline docs/topics/i18n/timezones.txt:694]: A link to the docs would be helpful

File: tests/utils_tests/test_timezone.py [hunk 5]
```diff
@@ -2,41 +2,58 @@
...
-    needs_zoneinfo = unittest.skip("Test requires zoneinfo")
+    needs_pytx = unittest.skip('Test requires pytz')
 else:
-    PARIS_ZI = zoneinfo.ZoneInfo('Europe/Paris')
-    PARIS_IMPLS = (CET, PARIS_ZI)
+    CET = pytz.timezone('Europe/Paris')
+    PARIS_IMPLS = (PARIS_ZI, CET)
 
-    def needs_zoneinfo(f):
+    def needs_pytx(f):
```
Comment (smithdc1) [inline tests/utils_tests/test_timezone.py:34]: Pytz?

Comment (carltongibson) [reply tests/utils_tests/test_timezone.py]: Nice. 😄 Thanks.

File: docs/releases/4.0.txt (hunk 1 above)
Comment (carltongibson) [reply docs/releases/4.0.txt:59]: Nice bit of French auto-correct for you there ⬆️ @aaugustin 😀

File: docs/topics/i18n/timezones.txt (hunk 4 above)
Comment (carltongibson) [reply docs/topics/i18n/timezones.txt:694]: The ``:func:`zoneinfo.available_timezones`` is linked  🤔

Comment (smithdc1) [reply docs/topics/i18n/timezones.txt:694]: Ah yes. 👍 I was focusing too much on the 'see the docs' part.

File: docs/releases/4.0.txt [hunk 6]
```diff
@@ -28,6 +28,47 @@ The Django 3.2.x series is the last to support Python 3.6 and 3.7.
...
+the move to ``zoneinfo`` should be largely transparent.
+
+If though you are you are working with non-UTC time zones, you will need to
+audit your code, since ``pytz`` and ``zoneinfo`` are not entirely equivalent:
+
+* The transitional :setting:`USE_DEPRECATED_PYTZ` setting is provided to
+  continue using ``pytz`` during the 4.x release cycle. This setting will be
+  removed in Django 5.0.
```
Comment (adamchainz) [inline docs/releases/4.0.txt:54]: I don't think this belongs in this list, which is meant to be a list of ways pytz and zoneinfo are not equivalent. Put it in a paragraph after the `pytz_deprecation_shim` one?

File: docs/topics/i18n/timezones.txt [hunk 7]
```diff
@@ -32,15 +31,20 @@ True <USE_TZ>` in your settings file.
 
     In Django 5.0, time zone support will be enabled by default.
 
-By default, time zone support uses pytz_, which is installed when you install
-Django; Django also supports the use of other time zone implementations like
-:mod:`zoneinfo` by passing :class:`~datetime.tzinfo` objects directly to
-functions in :mod:`django.utils.timezone`.
+Time zone support uses :mod:`zoneinfo`, which is part of the Python standard
+library from Python 3.9. A backport is installed when you install Django if you
+are using Python 3.8.
```
Comment (adamchainz) [inline docs/topics/i18n/timezones.txt:36]:
```suggestion
library from Python 3.9. The backport package ``backports.zoneinfo`` is
installed when you install Django on Python 3.8.
```

Mentioning the package name improves discoverability.

File: docs/topics/i18n/timezones.txt [hunk 8]
```diff
@@ -198,12 +202,15 @@ Create a view that can set the current timezone::
 
     from django.shortcuts import redirect, render
 
+    # Prepare a list of common timezone choices you wish to offer.
+    common_timezones = [...]
+
     def set_timezone(request):
         if request.method == 'POST':
             request.session['django_timezone'] = request.POST['timezone']
```
Comment (adamchainz) [inline docs/topics/i18n/timezones.txt:210]: this view should validate the timezone is in `common_timezones`

File: docs/topics/i18n/timezones.txt [hunk 9]
```diff
@@ -198,12 +202,15 @@ Create a view that can set the current timezone::
 
     from django.shortcuts import redirect, render
 
+    # Prepare a list of common timezone choices you wish to offer.
+    common_timezones = [...]
```
Comment (adamchainz) [inline docs/topics/i18n/timezones.txt:206]: include a concrete example - afaiu this should be a list of strings, not a list of `ZoneInfo`, which some might interpret the comment as implying

File: django/utils/timezone.py [hunk 10]
```diff
@@ -229,8 +256,16 @@ def is_naive(value):
...
+def make_aware(value, timezone=None, is_dst=IS_DST_PASSED):
     """Make a naive datetime.datetime in a given time zone aware."""
+    if is_dst is IS_DST_PASSED:
+        is_dst = None
+    else:
+        warnings.warn(
+            'The is_dst argument to make_aware() is deprecated as it has no '
+            'effect with zoneinfo time zones.',
+            RemovedInDjango50Warning
+        )
```
[the remark this answers is not in the document]
Comment (felixxm) [reply django/utils/timezone.py:268]: Please add a test for this warning.

**Review by felixxm — COMMENTED** [review]
@carltongibson Many thanks :+1:

File: django/db/migrations/serializer.py [hunk 11]
```diff
@@ -67,7 +67,14 @@ def serialize(self):
         imports = ["import datetime"]
         if self.value.tzinfo is not None:
             imports.append("from django.utils.timezone import utc")
-        return repr(self.value).replace('<UTC>', 'utc'), set(imports)
+        # Serialize, replacing tzinfo implementation details:
+        serialized = (
+            repr(self.value)
+            .replace('backports.', '')  # zoneinfo for PY38
+            .replace("zoneinfo.ZoneInfo(key='UTC')", 'utc')
```
[the remark this answers is not in the document]
Comment (carltongibson) [reply django/db/migrations/serializer.py:74]: This is covered by `migrations.test_writer.WriterTests.test_serialize_datetime` which fails with `name 'backports' is not defined` on PY38 and `name 'zoneinfo' is not defined` on PY39.

File: tests/timezones/tests.py [hunk 12]
```diff
@@ -362,6 +363,25 @@ def test_query_filter_with_pytz_timezones(self):
...
+            tests += [
+                (True, None, pytz.BaseTzInfo),
+                (True, 'Africa/Nairobi', pytz.BaseTzInfo),
+            ]
+        for use_pytz, connection_tz, expected_type in tests:
+            with self.subTest(use_pytz=use_pytz, connection_tz=connection_tz):
+                connection.timezone.clear_cache()
+                with override_settings(USE_DEPRECATED_PYTZ=use_pytz), \
+                     override_database_connection_timezone(connection_tz):
+                    self.assertIsInstance(connection.timezone, expected_type)
```
[the remark this answers is not in the document]
Comment (carltongibson) [reply tests/timezones/tests.py:382]: Don't like the continuation, hey? 😜

File: docs/topics/i18n/timezones.txt [hunk 13]
```diff
@@ -231,9 +238,8 @@ When you enable time zone support, Django interprets datetimes entered in
 forms in the :ref:`current time zone <default-current-time-zone>` and returns
 aware datetime objects in ``cleaned_data``.
 
-If the current time zone raises an exception for datetimes that don't exist or
-are ambiguous because they fall in a DST transition (the timezones provided by
-pytz_ do this), such datetimes will be reported as invalid values.
+Converted datetimes that don't exist or are ambiguous because they fall in a
```
Comment (pganssle) [inline docs/topics/i18n/timezones.txt:241]: I can see from the tests that this is the case (except for one situation, which is marked as a bug 😅), but is there some advice for how people are supposed to handle this situation? Probably non-existent times won't be an issue, but ambiguous times are real, just under-specified.

Is the idea that for a form in `Europe/Paris`, you'd want to specify, e.g. `'2011-10-30 02:30:00'` as either `'2011-10-30 02:30:00+01:00'` or `'2011-10-30 02:30:00+02:00'`? It seems to defeat the purpose of specifying "local time + zone" if the end user needs to know the offset in a zone at a specific time. Presumably the best thing to do would be to allow the person creating the form (possibly even via user input) to specify a strategy for what to do in the case of ambiguous datetimes?

I suppose this is just an aside, though, since the actual underlying behavior is not changed in this particular PR.

**Review by pganssle — COMMENTED** [review]
Sorry for the long delay in reviewing this. I have a few comments, but overall I think this looks _really_ good. I spent a bunch of time working out alternate strategies in a few places until I either realized what you were trying to do or found that you had anticipated some problem that I hadn't considered at first, so huzzah for that.

The biggest remaining open question for me is how to represent UTC. I still firmly believe that UTC should be `datetime.timezone.utc`. I think `zoneinfo.ZoneInfo('UTC')` is a weird anomaly, and if you have any sort of function that takes strings and gives back time zones, you should probably special-case `'UTC'` and `'utc'` to return `datetime.timezone.utc` (and to the extent that you have type hints, these functions should return `datetime.tzinfo` and *not* `zoneinfo.ZoneInfo`).

File: django/utils/timezone.py [hunk 14]
```diff
@@ -20,14 +27,26 @@
     'is_aware', 'is_naive', 'make_aware', 'make_naive',
 ]
 
+# RemovedInDjango50Warning
+IS_DST_PASSED = object()
```
[the remark this answers is not in the document]
Comment (carltongibson) [reply django/utils/timezone.py:31]: Just about to rename this now 😀

**Comment by carltongibson** [comment]
Hey @pganssle — thanks so much for taking the time to input again here! I'll address your various comments tomorrow now. 😴 

Just on this one: 

> The biggest remaining open question for me is how to represent UTC. I still firmly believe that UTC should be datetime.timezone.utc...

OK, so I was leaning towards going with this, on the basis of your recommendation basically, but I worried that we'd end up with two types in play. Most would be `ZoneInfo`, but then there'd be this one case where it was a `datetime.timezone`. 

Maybe it's because I spend too long doing triage but, I just have this nagging feeling that it's asking for trouble. 

> ... and to the extent that you have type hints, these functions should return `datetime.tzinfo` and not `zoneinfo.ZoneInfo`

Nice. I mentioned to Mariusz just today that we didn't want a type hint of `Union[datetime.timezone, zoneinfo.ZoneInfo]` — but you're right, we could (should!) use the superclass. 

I think users, though, won't restrict themselves to the public interface of `tzinfo` — they'll use the whole API (whether public or not) of `ZoneInfo`, except they'll have to add code to handle the deviant case. (We **know** the DRF test suite **already** depends on a comparable type being used in both places. — we could break that but it shows the point. It's Hyrum's Law right.) 

Using `datetime.timezone.utc` would be cleaner now —— that whole `__getattr__` thing is a bit 😬 — but from Django 5.0 knowing that we always have a `ZoneInfo` would be one less thing to worry about. (I'm sure there's a technical term in Category Theory for that 🤔)

Or that was my thinking... That you feel strongly is (still) quite weighty in the other way. We've got a few days. We can think again. 🧐 

Thanks!

**Comment by carltongibson** [comment]
CI is failing because I need to rebase after #14437, which I'll do tomorrow.

File: docs/releases/4.0.txt [hunk 15]
```diff
@@ -28,6 +28,47 @@ The Django 3.2.x series is the last to support Python 3.6 and 3.7.
...
+deprecated and will be removed in Django 5.0.
+
+:mod:`zoneinfo` is part of the Python standard library from Python 3.9. The
+``backports.zoneinfo`` package is automatically installed alongside Django if
+you are using Python 3.8.
+
+To the extent that you are always operating on aware datetimes with UTC, then
+the move to ``zoneinfo`` should be largely transparent.
+
+If though you are you are working with non-UTC time zones, you will need to
```
[the remark this answers is not in the document]
Comment (ngnpope) [reply docs/releases/4.0.txt:49]: I'd just change this to `However, if…` to avoid breaking the flow so many times.

File: django/utils/timezone.py [hunk 16]
```diff
@@ -20,14 +27,26 @@
...
-utc = pytz.utc
+# RemovedInDjango50Warning
+# This code has two purposes.
+# * Adjust the `utc` constant based on the `USE_DEPRECATED_PYTZ` setting.
+# * Defer that until used, since this module can be imported early in the
+#   bootstrap process before the settings are configured.
+#
+# For Django 5.0 Replace with:
+# utc = zoneinfo.ZoneInfo('UTC')
+def __getattr__(name):
```
[the remark this answers is not in the document]
Comment (ngnpope) [reply django/utils/timezone.py:42]: Good shout, thanks! Forgot about `__dir__()` when I proposed this.

File: docs/topics/i18n/timezones.txt [hunk 17]
```diff
@@ -19,10 +19,9 @@ practice to store data in UTC in your database. The main reason is daylight
 saving time (DST). Many countries have a system of DST, where clocks are moved
 forward in spring and backward in autumn. If you're working in local time,
 you're likely to encounter errors twice a year, when the transitions happen.
-(The pytz_ documentation discusses `these issues`_ in greater detail.) This
-probably doesn't matter for your blog, but it's a problem if you over-bill or
-under-bill your customers by one hour, twice a year, every year. The solution
-to this problem is to use UTC in the code and use local time only when
+This probably doesn't matter for your blog, but it's a problem if you over-bill
+or under-bill your customers by one hour, twice a year, every year. The
```
[the remark this answers is not in the document]
Comment (aaugustin) [reply docs/topics/i18n/timezones.txt:23]: I used to over-hyphen a lot a decade ago when I wrote this. Not my native language, sorry :-(

Interestingly, Firefox' spell checker just auto-corrected overhyphen to over-hyphen and autocorrected to auto-corrected. This is getting very meta ;-)

**Comment by aaugustin** [comment]
@carltongibson asked if I had any input to give on `django.utils.timezone.utc`. I'll do my best to help break the stalemate that I saw slowly solidifying in this discussion. Unfortunately you get another novel to digest...

While I was the timezones guy in Django-land for a few years, I haven't followed closely the storing around zoneinfo. Please keep in mind that @pganssle knows more than me about zoneinfo and everything around it.

----

So, here are the facts that I gathered — and some Opinions around those facts.

`django.utils.timezone.utc` started its life as an alias for `pytz.utc` when timezone support was enabled (`USE_TZ = True`) and a basic implementation copied form Python's docs otherwise ([source](https://github.com/django/django/commit/9b1cb755a28f020e27d4268c214b25315d4de42e#diff-698cf55de9163a8fc8381fc3f0320d8336203c45e03306f497c67495bffce40bR87)).

pytz' docs suggest that its implementation is exactly the same as what Django does when `USE_TZ = False`:

> The pytz package encourages using UTC for internal timezone representation by including a special UTC implementation based on the standard Python reference implementation in the Python documentation.

Why did I choose to use pytz when it was available, then, rather than always using the version copy-pasted from the docs?

Maybe I believed pytz was faster, in which case I certainly wanted to use it when available. Alternatively, as much as possible, I wanted to use an implementation maintained by someone who knew what they were doing better than I did. However, I still had to provide a fallback when pytz wasn't available because we wouldn't add dependencies to Django back then.

If the stdlib provided an UTC definition, I'm sure I'd use it instead of rolling our own. I'm pretty sure I wasn't doing this to support calling pytz' `normalize()` method on `django.utils.timezone.utc`.

A few years later, Django finally accepts dependencies, adds a dependency on pytz, and removes the fallback copy-pasted from Python's docs ([source](https://github.com/django/django/commit/414ad25b090a63eaaf297b1164c8f7d814a710a2#diff-698cf55de9163a8fc8381fc3f0320d8336203c45e03306f497c67495bffce40bR53)). This supports my theory that the copy-pasted implementation was just a fallback. I reviewed the related discussion; at no point we discuss the `utc` definition.

At this point, the only reason why `django.utils.timezone.utc` exists at all is because the stdlib didn't provide it and we don't want to tell everyone to write `pytz.utc` in their code. Then the stdlib [provides it](https://docs.python.org/3/library/datetime.html#datetime.timezone.utc) in Python 3.5.

So we [try to adopt it](https://github.com/django/django/commit/27ca5ce19f5f184018a61611c1bc319113b1d107#diff-698cf55de9163a8fc8381fc3f0320d8336203c45e03306f497c67495bffce40bR57) and then [rollback](https://github.com/django/django/commit/2ec151e35da93047acfeea1b79c27010f2cb8594#diff-698cf55de9163a8fc8381fc3f0320d8336203c45e03306f497c67495bffce40bR55) due to [this ticket](https://code.djangoproject.com/ticket/29514).

----

I have ambivalent feelings about that ticket. On one hand, the documentation for `django.utils.timezone.utc` just says "that's a tzinfo object representing UTC". On the other hand, rejecting that ticket as invalid doesn't bode well for the Principle of Least Astonishment — it's somewhat reasonable to expect the UTC timezone to be a singleton.

Overall, I'd be OK with telling DRF to write `timezone.TimeZone("UTC")` instead of `django.utils.timezone.utc` in this case. If they wrote the test in a clean room in a world where `timezone.get_default_timezone()` returns `timezone.TimeZone("UTC")`, they'd just use that in the test and call it a day.

----

I think my favorite option would be to deprecate `django.utils.timezone.utc` in favor of `datetime.timezone.utc`. Then we get rid of an API that no longer serves any purpose now that Python gives as a built-in, always available version of the UTC timezone (better late than never!)

In a world where `django.utils.timezone.utc` no longer exists, it's even more obvious that DRF writes `timezone.TimeZone("UTC")` — and, if they write `datetime.timezone.utc` for any reason and get confused, we can point them to https://bugs.python.org/ :-)

The obvious downside is code churn. Keep in mind that a large chunk of our userbase hates my liberal interpretation of the backwards-compatibility policy (they still remember Django 1.7).

----

Long story short, it is quite unfortunate that the history of the Python stdlib provides two incompatible versions of the UTC tzinfo. If only we had design principles such as "There should be one-- and preferably only one --obvious way to do it." we wouldn't be having this discussion :trollface: 

We have to accept that we inherit this problem from Python. If we pretend that we can have only `datetime.timezone.utc` or `zoneinfo.ZoneInfo('UTC')`, we're lying to ourselves. Better admit that both can happen and equip our users to handle that.

----

I don't see a major reason to use `datetime.timezone.utc` rather than `zoneinfo.ZoneInfo('UTC')` or vice versa. Anyway, in some cases we're going to get one and in other cases we're going to get the other.

A lot of aware timezones coming from the outside world are likely to use `datetime.timezone.utc`. I think we'd like to minimize the impedance mismatch there so I'm in favor of `datetime.timezone.utc`.

"Simplest option wins" is another argument in favor of `datetime.timezone.utc`.

Should we special-case `get_default/currrent_timezone()` to return `datetime.timezone.utc` when the default or current timezone is defined as UTC? No, I think this is a step too far.

----

**tl;dr**

- consider deprecating `django.utils.timezone.utc` in favor of `datetime.timezone.utc`; failing that go for an alias for `datetime.timezone.utc`
- probably you still need to keep `pytz.utc` when `USE_DEPRECATED_PYTZ = True` — the price to pay for the size of our userbase
- consider adding a blurb to the docs about `datetime.timezone.utc` and `zoneinfo.ZoneInfo('UTC')` being equivalent but not the same; if we have a timezone FAQ that would be a good place
- the timezone comparison story becomes "look what tzinfo you get and compare against that"

I hope this helps 🤞

**Comment by adamchainz** [comment]
> consider deprecating `django.utils.timezone.utc` in favor of `datetime.timezone.utc`; failing that go for an alias for `datetime.timezone.utc`

I'm in favour of deprecating. Django shouldn't have aliases to the standard library.

(Plus this kind of "import x -> import y" deprecation is pretty easy to add a fixer for in ma new projec' [django-upgrade](https://github.com/adamchainz/django-upgrade/).)

**Comment by ngnpope** [comment]
+1 to deprecating `django.utils.timezone.utc`. (Although we need to think on how that affects migrations.)

**Comment by carltongibson** [comment]
Hey @aaugustin -- thanks, yes great. 

> We have to accept that we inherit this problem from Python. If we pretend that we can have only datetime.timezone.utc or zoneinfo.ZoneInfo('UTC'), we're lying to ourselves. Better admit that both can happen and equip our users to handle that.

OK, let's do it. 👍 

I have to think about deprecating the utils alias. It might be we could bump that to 4.1 🤔

File: django/utils/timezone.py [hunk 18]
```diff
@@ -257,11 +292,25 @@ def make_naive(value, timezone=None):
 
 def _is_pytz_zone(tz):
     """Checks if a zone is a pytz zone."""
+    # Try import rather than checking settings.USE_DEPRECATED_PYTZ to *allow*
+    # manually passing a pytz timezone, which some of the test cases (at least)
+    # rely on.
+    try:
+        import pytz
```
[the remark this answers is not in the document]
Comment (carltongibson) [reply django/utils/timezone.py:300]: That's super Paul. Thanks!

**Comment by aaugustin** [comment]
> we need to think on how that affects migrations.

We could dynamically inject it only in the context of running migrations... This would be a questionnable design though ;-) Or just bite the bullet and tell people to run a "replace-all" on their migrations.

> I have to think about deprecating the utils alias. It might be we could bump that to 4.1 🤔

I don't remember how the deprecation process works these days. Ideally you want it gone from 5.0, together with `USE_DEPRECATED_PYTZ`.

File: django/db/migrations/serializer.py [hunk 19]
```diff
@@ -67,7 +67,14 @@ def serialize(self):
         imports = ["import datetime"]
         if self.value.tzinfo is not None:
             imports.append("from django.utils.timezone import utc")
-        return repr(self.value).replace('<UTC>', 'utc'), set(imports)
+        # Serialize, replacing tzinfo implementation details:
+        serialized = (
```
[the remark this answers is not in the document]
Comment (carltongibson) [reply django/db/migrations/serializer.py:71]: Switching to `datetime.timezone.utc`, this looks to become moot... See https://github.com/django/django/pull/14848/commits/f3b06db3954bd3a9d452e1eeb4b4862504a98c48

File: django/db/migrations/serializer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (carltongibson) [reply django/db/migrations/serializer.py]: This looks like it's made moot by switching to `datetime.timezone.utc`  See https://github.com/django/django/pull/14848/commits/f3b06db3954bd3a9d452e1eeb4b4862504a98c48

File: django/utils/timezone.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (carltongibson) [reply django/utils/timezone.py]: https://github.com/django/django/pull/14848/commits/f3b06db3954bd3a9d452e1eeb4b4862504a98c48 switched to use `datetime.timezone.utc` so this drops away. (Thanks both!)

File: django/db/migrations/serializer.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (carltongibson) [reply django/db/migrations/serializer.py:71]: Deprecating the `utils.timezone.utc` alias will let us use `datetime.timezone.utc` directly in the migrations here.

File: django/utils/timezone.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (carltongibson) [reply django/utils/timezone.py]: Although, no it doesn't  😀 — I think I'm gonna need this to deprecate `.utc`.

**Comment by carltongibson** [comment]
_Getting There™_. 

I've addressed all review comments so far **except**:

* The docs suggestions/changes, which I'll do tomorrow morning on a fresh head. I'll then look to squash down, so if you have any lingering thoughts/suggestions, now would be good for those. 😀 
* The proposed deprecation of `django.utils.timezone.utc`.

On the deprecation, I'd like to leave that to a separate PR, regardless of what we want to do. (It's _everywhere_ and I'd just like to _measure twice_…) 

> I don't remember how the deprecation process works these days. Ideally you want it gone from 5.0...

If we were to deprecate in 4.1, we can remove in 5.0. [Policy](https://docs.djangoproject.com/en/3.2/internals/release-process#s-internal-release-deprecation-policy): 

> Deprecated features will be removed in the B.0 release, or B.1 for features deprecated in the last A.x feature release...

Thanks for all the help all. 🎁

**Comment by felixxm** [comment]
> I don't remember how the deprecation process works these days. Ideally you want it gone from 5.0, together with `USE_DEPRECATED_PYTZ`.

Things deprecated in Django 4.0 and 4.1 will be removed in 5.0.

**Comment by aaugustin** [comment]
OK thanks!

File: docs/topics/i18n/timezones.txt (hunk 13 above)
Comment (carltongibson) [reply docs/topics/i18n/timezones.txt:235]: I'm going to pass on this point for now. We don't really have a story, except earlier on there's a _In practice, this is rarely an issue._ … 

I will ponder it going forward though. 

Thanks.

File: docs/topics/i18n/timezones.txt (hunk 8 above)
Comment (carltongibson) [reply docs/topics/i18n/timezones.txt:204]: At the beginning of this section we have: 

> Here's an example that stores the current timezone in the session. (It skips
error handling entirely for the sake of simplicity.)

So I'm going to take error handling as out of scope (unless the example gets rewritten separately)

**Comment by carltongibson** [comment]
Hey @pganssle @aaugustin — I've addressed all the comments thus far, rebased, squashed and so on. Can I ask if you have any further comments for now? (😬) 

Thanks!

**Comment by aaugustin** [comment]
I am officially out of comments :-)

File: django/db/migrations/serializer.py [hunk 20]
```diff
@@ -67,7 +67,12 @@ def serialize(self):
...
+            repr(self.value)
+            .replace('datetime.timezone.utc', 'utc')
+        )
+        return serialized, set(imports)
```
Comment (felixxm) [inline django/db/migrations/serializer.py:75]: Can we restore the previous "style"?
```suggestion
        return repr(self.value).replace('datetime.timezone.utc', 'utc'), set(imports)
```

File: django/utils/timezone.py [hunk 21]
```diff
@@ -229,8 +241,17 @@ def is_naive(value):
...
+def make_aware(value, timezone=None, is_dst=NOT_PASSED):
     """Make a naive datetime.datetime in a given time zone aware."""
+    if is_dst is NOT_PASSED:
+        is_dst = None
+    else:
+        warnings.warn(
+            'The is_dst argument to make_aware(), used by the Trunc() '
+            'database functions and QuerySet.datetimes(), is deprecated as it '
+            'has no effect with zoneinfo time zones.',
+            RemovedInDjango50Warning
```
Comment (felixxm) [inline django/utils/timezone.py:253]:
```suggestion
            RemovedInDjango50Warning,
```

File: django/db/models/functions/datetime.py [hunk 22]
```diff
@@ -264,7 +264,7 @@ def convert_value(self, value, expression, connection):
 
 class Trunc(TruncBase):
 
-    def __init__(self, expression, kind, output_field=None, tzinfo=None, is_dst=None, **extra):
+    def __init__(self, expression, kind, output_field=None, tzinfo=None, is_dst=timezone.NOT_PASSED, **extra):
```
Comment (felixxm) [inline django/db/models/functions/datetime.py:267]: We can miss removing `is_dst` when bootstrapping Django 5.0, maybe a small comment.
```suggestion
    # RemovedInDjango50Warning: when the deprecation ends, remove is_dst
    # argument.
    def __init__(self, expression, kind, output_field=None, tzinfo=None, is_dst=timezone.NOT_PASSED, **extra):
```

File: django/db/models/functions/datetime.py [hunk 23]
```diff
@@ -188,7 +188,7 @@ class TruncBase(TimezoneMixin, Transform):
     kind = None
     tzinfo = None
 
-    def __init__(self, expression, output_field=None, tzinfo=None, is_dst=None, **extra):
+    def __init__(self, expression, output_field=None, tzinfo=None, is_dst=timezone.NOT_PASSED, **extra):
```
Comment (felixxm) [inline django/db/models/functions/datetime.py:191]:
```suggestion
    # RemovedInDjango50Warning: when the deprecation ends, remove is_dst
    # argument.
    def __init__(self, expression, output_field=None, tzinfo=None, is_dst=timezone.NOT_PASSED, **extra):
```

File: django/db/models/query.py [hunk 24]
```diff
@@ -916,7 +916,7 @@ def dates(self, field_name, kind, order='ASC'):
             'datefield', flat=True
         ).distinct().filter(plain_field__isnull=False).order_by(('-' if order == 'DESC' else '') + 'datefield')
 
-    def datetimes(self, field_name, kind, order='ASC', tzinfo=None, is_dst=None):
+    def datetimes(self, field_name, kind, order='ASC', tzinfo=None, is_dst=timezone.NOT_PASSED):
```
Comment (felixxm) [inline django/db/models/query.py:919]:
```suggestion
    # RemovedInDjango50Warning: when the deprecation ends, remove is_dst
    # argument.
    def datetimes(self, field_name, kind, order='ASC', tzinfo=None, is_dst=timezone.NOT_PASSED):
```

File: docs/ref/utils.txt [hunk 25]
```diff
@@ -960,6 +960,11 @@ appropriate entities.
     The ``is_dst`` parameter has no effect when using non-``pytz`` timezone
     implementations.
 
+    .. deprecated:: 4.0
```
Comment (felixxm) [inline docs/ref/utils.txt:963]: Maybe we should move the entire paragraph about `pytz` under `deprecated` :thinking: Starting with ```When using ``pytz``... ```

File: tests/datetimes/tests.py [hunk 26]
```diff
@@ -91,7 +96,9 @@ def test_21432(self):
         qs = Article.objects.datetimes('pub_date', 'second')
         self.assertEqual(qs[0], now)
 
-    @override_settings(USE_TZ=True, TIME_ZONE='UTC')
+    @unittest.skipUnless(pytz is not None, "Test needs pytz")
```
Comment (felixxm) [inline tests/datetimes/tests.py:99]:
```suggestion
    @unittest.skipUnless(pytz is not None, 'Test requires pytz')
```

**Review by felixxm — APPROVED** [review]
@carltongibson Thanks for all your massive efforts :rocket: 

@aaugustin @pganssle Thanks for the extensive reviews and discussion :star2:

**Comment by aaugustin** [comment]
Hey @carltongibson - thank you so much for getting this PR to the finish line - it wasn't an easy one!
